spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java

```diff
@@ -736,58 +769,67 @@ protected void expireGroup(Object correlationKey, MessageGroup group) {
 				this.logger.debug("Prematurely releasing partially complete group with key ["
 						+ correlationKey + "] to: " + getOutputChannel());
 			}
-			completeGroup(correlationKey, group);
+			completeGroup(correlationKey, group, lock);
 		}
 		else {
 			if (this.logger.isDebugEnabled()) {
 				this.logger.debug("Discarding messages of partially complete group with key ["
 						+ correlationKey + "] to: "
 						+ (this.discardChannelName != null ? this.discardChannelName : this.discardChannel));
 			}
-			for (Message<?> message : group.getMessages()) {
-				discardMessage(message);
+			if (this.releaseLockBeforeSend) {
+				lock.unlock();
 			}
+			group.getMessages().forEach(this::discardMessage);
 		}
 		if (this.applicationEventPublisher != null) {
 			this.applicationEventPublisher.publishEvent(new MessageGroupExpiredEvent(this, correlationKey, group
 					.size(), new Date(group.getLastModified()), new Date(), !this.sendPartialResultOnExpiry));
 		}
 	}
 
-	protected void completeGroup(Object correlationKey, MessageGroup group) {
+	protected void completeGroup(Object correlationKey, MessageGroup group, Lock lock) {
 		Message<?> first = null;
 		if (group != null) {
 			first = group.getOne();
 		}
-		completeGroup(first, correlationKey, group);
+		completeGroup(first, correlationKey, group, lock);
 	}
 
 	@SuppressWarnings("unchecked")
-	protected Collection<Message<?>> completeGroup(Message<?> message, Object correlationKey, MessageGroup group) {
-		if (this.logger.isDebugEnabled()) {
-			this.logger.debug("Completing group with correlationKey [" + correlationKey + "]");
-		}
-
-		Object result = this.outputProcessor.processMessageGroup(group);
+	protected Collection<Message<?>> completeGroup(Message<?> message, Object correlationKey, MessageGroup group,
+			Lock lock) {
 		Collection<Message<?>> partialSequence = null;
-		if (result instanceof Collection<?>) {
-			verifyResultCollectionConsistsOfMessages((Collection<?>) result);
-			partialSequence = (Collection<Message<?>>) result;
-		}
+		Object result = null;
+		try {
+			if (this.logger.isDebugEnabled()) {
+				this.logger.debug("Completing group with correlationKey [" + correlationKey + "]");
+			}
 
-		if (this.popSequence && partialSequence == null && !(result instanceof Message<?>)) {
-			AbstractIntegrationMessageBuilder<?> messageBuilder;
-			if (result instanceof AbstractIntegrationMessageBuilder<?>) {
-				messageBuilder = (AbstractIntegrationMessageBuilder<?>) result;
+			result = this.outputProcessor.processMessageGroup(group);
+			if (result instanceof Collection<?>) {
+				verifyResultCollectionConsistsOfMessages((Collection<?>) result);
+				partialSequence = (Collection<Message<?>>) result;
 			}
-			else {
-				messageBuilder = getMessageBuilderFactory()
-						.withPayload(result)
-						.copyHeaders(message.getHeaders());
+
+			if (this.popSequence && partialSequence == null && !(result instanceof Message<?>)) {
```

Comment:
Hm. the `popSequence` is there since `5.1`.
I don't think this change is going to be back-ported to `5.0` so easy...

Maybe as a separate PR?

Reply:
Or don't backport since there is a workaround?

Reply:
OK. Let's don't back-port!

Please, proceed then with all the required changes to reflect only `5.1.1` version.

Thanks